### ISSUE TITLE
`CBPS` method for `match_on` 

### ISSUE THREAD

**Issue by benthestatistician** [markmfredrickson/optmatch#121]
I heard from users who tried to call `fullmatch` on a [CBPS](https://cran.r-project.org/package=CBPS) (Covariate Balancing Propensity Score, a la Imai and Ratcovic [2014]).  They found it feasible, but with a bit of doing; see forwarded message in comment below.  The description of their workaround suggests that it ought to be possible to encode these steps in `fullmatch.CBPS` and/or `pairmatch.CBPS` methods.  (Although I didn't have time to investigate further, I did check and confirm that the main output of the package is an S3 object called "CBPS".)  

If someone out there in userland has a hankering to help with this, please don't hesitate to contribute a patch to optmatch, ideally following [github standard fork-and-pull model](https://gist.github.com/Chaser324/ce0505fbed06b947d962).

**Comment by benthestatistician** [comment]
Here's that user email:

> While we have used with no problem "glm" to feed match_on and then fullmatch, we
encountered some problems when we attempted to use CBPS (R program by Imai and
colleagues) to arrive a propensity scores.  Here we found that fall match would
not accept the CBPS results as it would the glm results.  We had to take two
additional steps.

> First, we had to add names to the fitted.values and to the exposure variable
as follows.

```{r}
#32 is the nobs of psm in this example

names(psm$y)<-c(1:32)

#Next the fitted values (pscores)

names(psm$fitted.values)<-c(1:32)
```


This step was necessary even for the nuclearplants dataset.

Second, once we went beyond nuclearplants, we had to
cbind the fitted.values and the y vectors onto the original.data
so that "fullmatch" in its "data=original.data" statement would recognize the
dataset.   Again, this issue did not arise in nuclearplants but in
other datasets.

**Comment by benthestatistician** [comment]
Here's an example of the procedure this issue would simplify. (From Russell Localio, with minor adjustments.)

```{r}
library('CBPS')
library('optmatch')
#glm code
psm<- glm(pr ~ date+t1+t2+cap+ne+ct+bw+cum.n, family=binomial, data=nuclearplants)

#CBPS in place of glm
psm<- CBPS(pr ~ date+t1+t2+cap+ne+ct+bw+cum.n, ATT=0, data=nuclearplants)


# Step(2): with pscores (psm), make the distance matrix
#must add names to psm because CBPS does not name vector the way that glm does
#Crucial step 
#
#Use names function
# First the treatment indicator (outcome of CBPS)
#32 is the nobs of psm in this example

names(psm$y)<-rownames(nuclearplants)

#Next the fitted values (pscores)

names(psm$fitted.values)<-rownames(nuclearplants)
#Note syntax for match_on now requires: 
psm.dist<-match_on(x=psm$fitted.values, z=psm$y, data=nuclearplants)
#
# Step(3): Full match based on the distance matrix  
#
plant.match<-fullmatch(psm.dist,data=nuclearplants,min=2, max=3)

```

**Comment by benthestatistician** [comment]
Notes: 

1.  After re-familiarizing myself with the _current_ architecture of optmatch, I think the task is just to write a `match_on` method. Shouldn't also need `fullmatch` and/or `pairmatch` methods, as issue statement now suggests.  (I'll update the issue title after posting this.)
2. That `match_on` method could just wrap to `match_on.glm` after adding in a `linear.predictor` slot to the "CBPS" object.  (The CBPS class seems to inherit from glm; its lack of a `linear.predictor` seems to be only reason this doesn't work out of the box.)
3. I propose that the linear predictor of a CBPS just be the inverse logits of the fitted probabilities (its fitted values). 
4. There could well be users who would prefer to match on the fitted probabilities themselves.  I'd oppose having this as a default setting, but I wouldn't oppose allowing it, either for CBPS's or for ordinary glms.  We could add this as an option to `match_on.glm`. 

Comments appreciated.

**Comment by josherrickson** [comment]
The technical reason this doesn't work out of the box (ignoring the check for `linear.predictor` at the beginning of `match_on.glm`) is that `predict(CBPS(...))` returns `NULL`, which is breaking `scores`. 

Perhaps the lightest touch to the code would therefore be to add a special case for `CBPS` to `scores`? Such a special case already exists for `bigglm`.<sup>1</sup> 

I think we'd just need to replace the final call to `predict` with our own version.

It would also be nice to future-proof a bit, such that if `predict.CBPS` is supported, it prefers that over our method without us needing to release an update. Something like this?

```r
if (is(object, 'CBPS')) {
  if (is.null(pp <- eval(predict(object))) {
   return(<our version of predicted values>)
  } else {
   return(pp)
  }
}
```


1: A more formal solution would to make `scores` dispatch on object type, though of course, that's far more work.

**Comment by benthestatistician** [comment]
How about we just introduce our own `predict.CBPS`?  We wouldn't have to export it.

**Comment by josherrickson** [comment]
That would indeed be an even lighter touch. We should still build in future-proofing I think, similar to my code snippet above.

**Comment by benthestatistician** [comment]
The future-proofing is clever, but to my mind it introduces a source of irreproducibility into code scripts.   If the `CBPS` group should introduce a `predict` method for `CBPS`'s, then people who update `CBPS` will suddenly start getting different results out of optmatch.  If they updated their optmatch at the same time, e.g. in the midst of an update to a newer version of R itself, then it will seem to them like it was an optmatch related change.  And we'll be hard pressed to diagnose the issue if asked.

Particularly if this chain of events were to occur without the `CBPS` group notifying us about it.  On the other hand if they see that we've implemented our own `predict.CPBS` and decide they'd recommend something different than what we've done, then they can always contact us with a patch, or a request that we deprecate our method in favor of theirs.

**Comment by josherrickson** [comment]
I've committed a [branch](https://github.com/markmfredrickson/optmatch/tree/cbps) with a basic version of this implemented. @benthestatistician feel free to jump in and update as necessary. It's very bare-bones right now.

I explicitly did not use `predict.CBPS` to avoid some errors I was getting. I did not include any future proofing.

**Comment by josherrickson** [comment]
I did some testing by giving CBPS a `predict.CBPS` method, and saw no issues if we likewise call our function `predict.CBPS` and don't export it.

Just calling `predict(CBPS)` inside `scores` automatically calls our `optmatch:::predict.CBPS` regardless of the existence of `CBPS::predict.CBPS`. This means that if CBPS implements predict, we need only remove our version without touching `scores`.

Do we want some sort of message to users if this happens? Something like "Using internal version of predict.CBPS, check for an updated version of optmatch to use CBPS::predict.CBPS"?

**Comment by josherrickson** [comment]
Merged CBPS branch back into master. I think this is all set unless @benthestatistician wants to add the warning I mused about in my last comment.

**Comment by benthestatistician** [comment]
If CBPS were to implement a `predict` method, before we started using it I'd want to know that they and/or we had considered its appropriateness for use in matching, as opposed to some other context.  (Recall that CBPS is a package for propensity weighting, not propensity matching.) 

Nice work, @josherrickson.  Thanks for touching up my `plogis` call.

**Comment by benthestatistician** [comment]
`predict.glm` has an argument `type=c("link", "response", "terms")`.  On reflection I think `predict.CBPS` should have something similar, `type=c("link", "response")`.  The default, "link", would return what's returned now, whereas giving `type="response"` would return the `fitted.values` of the provided CBPS object.  This would give users a mechanism for overriding the optmatch default of logit-scale matching in favor of probability-scale matching, the same such mechanism as would be used with matching on an ordinary propensity score.
